src/mongocxx/instance.cpp

```diff
@@ -14,31 +14,73 @@
 
 #include <mongocxx/instance.hpp>
 
+#include <mutex>
+#include <utility>
+
 #include <bsoncxx/stdx/make_unique.hpp>
 
-#include <mongoc.h>
+#include <mongocxx/logger.hpp>
+#include <mongocxx/private/libmongoc.hpp>
+
+namespace mongocxx {
+MONGOCXX_INLINE_NAMESPACE_BEGIN
 
 namespace {
-void log_handler(mongoc_log_level_t, const char *, const char *, void *) {
+
+log_level convert_log_level(::mongoc_log_level_t mongoc_log_level) {
+    switch (mongoc_log_level) {
+        case MONGOC_LOG_LEVEL_ERROR:
+            return log_level::k_error;
+        case MONGOC_LOG_LEVEL_CRITICAL:
+            return log_level::k_critical;
+        case MONGOC_LOG_LEVEL_WARNING:
+            return log_level::k_warning;
+        case MONGOC_LOG_LEVEL_MESSAGE:
+            return log_level::k_message;
+        case MONGOC_LOG_LEVEL_INFO:
+            return log_level::k_info;
+        case MONGOC_LOG_LEVEL_DEBUG:
+            return log_level::k_debug;
+        case MONGOC_LOG_LEVEL_TRACE:
+            return log_level::k_trace;
+        default:
+            // TODO: MONGOCXX_UNREACHABLE (CXX-628)
+            std::abort();
+    }
+}
+
+void null_log_handler(::mongoc_log_level_t, const char *, const char *, void *) {
 }
+
+void user_log_handler(::mongoc_log_level_t mongoc_log_level, const char *log_domain,
+                      const char *message, void *user_data) {
+    (*static_cast<logger *>(user_data))(convert_log_level(mongoc_log_level),
+                                        stdx::string_view{log_domain}, stdx::string_view{message});
 }
 
-namespace mongocxx {
-MONGOCXX_INLINE_NAMESPACE_BEGIN
+}  // namespace
 
 class instance::impl {
    public:
-    impl() {
-        mongoc_init();
-        mongoc_log_set_handler(log_handler, nullptr);
+    impl(std::unique_ptr<logger> logger) : user_logger(std::move(logger)) {
+        libmongoc::init();
+        if (user_logger) {
+            libmongoc::log_set_handler(user_log_handler, user_logger.get());
+        }
     }
 
     ~impl() {
-        mongoc_cleanup();
+        libmongoc::cleanup();
```

Comment:
Presumably, cleanup can log. Are we assured that after calling cleanup, the C driver will make no more calls to the logger? Or should we explicitly set it back to the null log handler here, rather than leaving it with a dangling pointer to a defunct user logger when we leave this scope?